README.md
# mindfulnotifier

A mindfulness bell with configurable text notifications, schedules, and sound.

## About this app

This app is a simple mindfulness-based timer which displays a notification/reminder at user-defined intervals. It is another take on a
'Mindfulness Bell', with the addition of accompanying text.  The reminders are taken from a configurable list, and are chosen at random at a
selected interval. The reminder interval can either be periodic (in intervals down to a 15 minute granularity) or random (between a selected
range of minutes).

Some default reminders are provided as examples.
- You can add, edit, or remove these default reminders as you like.
- Reminders can be enabled or disabled.
- Reminders can be assigned a 'tag', and the reminder list can be filtered by tag.

There are 5 bells included, and you can also configure a custom bell from your phone's local storage.

This app works very well in conjunction with a smartwatch. In this mode you can also mute the bell to have silent mindfulness prompts
throughout your day.

The app is built using [Flutter](https://flutter.dev/).  It currently only runs on Android.


### Features:

- Configure either a periodic or random reminder interval.
    - e.g. hourly, or every 15 minutes, or randomly between 30 minutes and an hour
    - periodic intervals are aligned from the top of the hour
- Edit or add your own reminder messages:
    - Organize/group by tag
    - Enable/disable reminders
- Define a time range for daily quiet hours.
- Choose from a set of default bells, or provide your own.
- Separate controls for 'mute' and 'vibrate'.
- Choose android audio output channel for sound: notification (default), media, or alarm


### Permissions

This app uses the following permissions:

- `ACCESS_NETWORK_STATE`: Required by the library used to launch the github source code site
- `RECEIVE_BOOT_COMPLETED`: Required by the underlying alarm service to restart after reboot
- `WAKE_LOCK`: Required by the underlying alarm service to wake up at the scheduled time
- `USE_FULL_SCREEN_INTENT`: Required to show app in foreground from the reminder notification
- `VIBRATE`: Required to vibrate during notifications


### Notes on Application Settings in Android

#### Battery Optimization

Some phones will kill apps when running in the background after a relatively short period of time.  Unfortunately, if your phone is killing
the app then you will stop getting notifications.  If this is happening, you must disable any 'battery optimization' settings for this app.
This varies from phone-to-phone - generally you must go into the settings for the app on your phone and turn off any battery optimization.
See https://dontkillmyapp.com/ for your phone type and android version for more information.


### Bells

- Bell 1: is from user 'steaq' via [https://freesound.org/s/346328](https://freesound.org/s/346328/) and is licensed under the [Creative Commons 0 License][CC0].

- Bell 2: is from user 'itsallhappening' via [https://freesound.org/people/itsallhappening/sounds/48795](https://freesound.org/s/48795/) and is licensed under the [Creative Commons Sampling+ License][CCS]

- Bell 3: is from user 'suburban grilla' via [https://freesound.org/people/suburban%20grilla/sounds/2166](https://freesound.org/s/2166/) and is licensed under the [Creative Commons Sampling+ License][CCS]

- Bell 4: is from user 'kerri' via [https://freesound.org/people/kerri/sounds/27421/](https://freesound.org/people/kerri/sounds/27421/) and is licensed under the [Creative Commons Attribution License][CCA]

- Bell 5: is from user 'dobroide' via [https://freesound.org/people/dobroide/sounds/436976/](https://freesound.org/people/dobroide/sounds/436976/) and is licensed under the [Creative Commons Attribution License][CCA]

[CC0]: http://creativecommons.org/publicdomain/zero/1.0/   "Creative Commons 0 License"
[CCS]: http://creativecommons.org/licenses/sampling+/1.0/  "Creative Commons Sampling+ License"
[CCA]: https://creativecommons.org/licenses/by/3.0/        "Creative Commons Attribution License"


### Reminder Import/Export

This app provides basic backup and restore functionality for the customized reminder list. Reminders can exported to a file (in JSON
format), which can then be restored on this device, or another device. Note that if you ever uninstall the app, you will lose all of your
customized reminders, so you may want to take a backup before uninstalling.

#### Export:

On export, the app saves a file called `Mindful Notifier-reminders-<date>.json` in the app's external storage directory, which is located in
your devices internal storage in the folder `<Internal Storage>/Android/data/com.kmac5dev.mindfulnotifier/files`.  It also offers you a
chance to 'share' the file via the standard Android share menu (i.e. copy it to another location on your device, to Google Drive, to
Dropbox, etc).

The 'share' step is important if:
- You uninstall the app. The app's storage directory is automatically removed when you uninstall the app.
- You want to share the settings between devices.

You can also use a file browser to access the `<Internal Storage>/Android/data/com.kmac5dev.mindfulnotifier/files` location.

Note: the exported file is a 'json' text file. It is possible to edit the contents, but if you mess it up that's on you ;-).

#### Import:

The import operation allows you to choose an exported backup file using the system's file chooser. The reminders will be restored from this
file. You should then restart the app, and re-enable the reminder service.

> Note: Previous 'backup/restore' functionality used a different file format. If you have an older backup file (named like `Mindful
> Notifier-backup-<date>.json`), you can still import this file. The reminders will be pulled from it; however, the other settings in the
> file will not be restored.


## Support Statement

This software is posted in the spirit of open-source software. I have created this app to fulfill my personal requirements, and I am sharing
it in case others may find it useful as well.  That said, my time is limited, and my approach to support is very much that I am happy if it
works for me. I will try to help if you are having issues, but I just don't have the time to add every desired new feature or troubleshoot
issues on hardware environments that I don't have access to.

I will consider pull requests for new features, and will always welcome bug fixes!


## License

This project is licensed under the terms of the GNU General Public License v3.0.

<p align="center">
  <a href="https://play.google.com/store/apps/details?id=com.kmac5dev.mindfulnotifier"><img alt="Get it on Google Play" src="https://play.google.com/intl/en_us/badges/images/apps/en-play-badge-border.png" height="75px"/></a>
  <a href="https://f-droid.org/en/packages/com.kmac5dev.mindfulnotifier"><img alt="Get it on F-Droid" src="https://fdroid.gitlab.io/artwork/badge/get-it-on.png" height="80"/></a>
</p>


CHANGELOG.md
# 1.0.24
- Updated all dependencies and caught up with flutter changes
- Removed sort-by-enabled in reminders view

# 1.0.23
- Back button now minimizes the app instead of closing it.
  This is like the home button is pressed instead of back.

# 1.0.22
- Fix issue where new notification is generated instead
  of showing existing notification (after app is restarted).

# 1.0.21
- Fix UI re-init issue with recreating the initial notification
  when app is revisited after being killed (either on restart
  or when user uses the back button to exit instead of the
  home button)
    - If user presses the back button the app is killed (but
      the alarm service is still running, and notifications
      still occur).
    - This fixes the notification being re-raised/changed
      when the app is again started by the user

# 1.0.20
- Issue #35: fix service disabled on app open after close
    - Migrate from SharedPreferences to Hive for internal data
    - add nextAlarm field in db to track state for restart
- Update to android targetSdkVersion 30 (required for Play Store)
- Update internal libraries and dependencies
- Remove some test reminders that were mistakenly included in the
  last release. This affects new installs only - if you recently
  installed please manually delete any reminders that don't make
  sense to you!

# 1.0.19
- Issue #31/#36: Revisit sound modes for different channels
- Summary:
    - App mute is enabled: always mute
    - For notification channel: honour the phone silent/vibrate-only setting
    - For media and alarm channels, allow sound unless DND is enabled

# 1.0.18
- Issue #32: Fix crash when reminder created with empty string

# 1.0.17
- Issue #32: Fix crash on earlier android versions caused by flutter_native_timezone

# 1.0.16
- 'Bell' configuration menu is now called 'Sound'
- Add ability to select media channel for notification sound (Issue 31):
    - notification (default, as in previous versions)
    - media
    - alarm

# 1.0.15
- Issue 26: allow to select from existing tags on add/edit
    - Can also add new tag via manual text
- Issue 25: Increase maximum characters per reminder
    - Increased from 256 to 4096 characters
    - Reminders are truncated in notifications. Will be
      shown in full on main screen via scrolling
- Issue 24: slight change to randomize function

# 1.0.14
- Rework reminder list UI and features for Issue #15 (tagging) and Issue #20
    - Ability to enable/disable individual reminders
    - Supports simple 'tag' field to categorize groups of reminders
    - Reminder list can be filtered by tag
    - Reminder list is always automatically sorted:
        - Alphabetical
        - Disabled reminders are moved to end of list
- Backup/Restore changed to Reminder Export/Import:
    - No longer does full backup/restore of all settings
        - NOTE: Existing backup files will still work - however, only the reminders will be restored.
    - Core function is changed to provide only reminder export/import
    - Reminder list export/import via JSON file format.
        - See: Issue #20 (simpler file format for reminder sharing)
    - New 'Merge' when importing reminder list:
        - 'Merge' will preserve existing reminders in list; only new reminders are added
        - 'Replace' completely replaces existing reminders list
- Fix issue #11, issue when updating schedule min/max values

# 1.0.13
- Issue #21 - ensure silent notifications in Do Not Disturb
- Issue #16 - add sharing option (long press on main screen to share)

# 1.0.12
- Issue 18 - fix periodic scheduling > 1 hr
- Issue 17 - add setting to hide next reminder info

# 1.0.11
- External storage changes (required by Google):
  - Remove android:requestLegacyExternalStorage="true" to conform to Android 11
  - Remove android.permission.READ_EXTERNAL_STORAGE, android.permission.WRITE_EXTERNAL_STORAGE permissions
- Backup/Restore changes (driven by the above):
  - Backup/Restore now use the application-specific external directory
    - Also given option to share the backup file using Android share facility
      - Share to local file system or cloud service, etc
      - This should be done to ensure the backup is preserved if you ever uninstall the app

# 1.0.10
- More fixes for move to data store in alarm/scheduler isolate
- Issue #9: fix dark color scheme dialogs
- Issue #6: fixes for backup/restore handling

# 1.0.9
- Update internal libraries
- Move all datastore writes into the alarm/scheduler isolate
  - Should fix issues with initial consistency of data

# 1.0.8
- Fix Issue #7 - fix bug in quiet hours end

# 1.0.7
- Remove quiet hours end callback

# 1.0.6
- Fixes after testing on Android 10 / LineageOS (Issue #2)
  - purely alarm driven
  - tweaking notifications which happen after main UI is killed

# 1.0.5
- cleanup/document permissions

# 1.0.4
- config changes required for publishing to fdroid

# 1.0.1
- minor code cleanup
- publishing to fdroid

# 1.0.0
- initial feature complete
- published to Google Play Store

# 0.0.1
- initial code


lib/screens/about.dart
// ignore_for_file: prefer_const_constructors

import 'package:flutter/material.dart';
import 'package:flutter_markdown/flutter_markdown.dart';
import 'package:url_launcher/url_launcher.dart';
import 'package:mindfulnotifier/components/constants.dart';
import 'package:mindfulnotifier/components/logging.dart';

var logger = createLogger('about');

const markdownData = """
## About this app

This app is a simple mindfulness-based timer which displays a
notification/reminder at user-defined intervals. It is another take on a
'Mindfulness Bell', with some accompanying text.  Reminders are taken from a
configurable list, and can be configured either for periodic or random
intervals.

Some default reminders are provided as examples. You can add, edit, or remove
these default reminders as you like. Reminders can be enabled/disabled and can
also be given a tag for easier grouping and browsing.

There are 5 bells included, and you can also configure a custom bell from your
phone's local storage.

This app works very well in conjunction with a smartwatch. In this mode you can
also mute the bell to have silent mindfulness prompts throughout your day.

The app is built using [Flutter](https://flutter.dev/).


### Features:

Main Features:

- Configure either a periodic or random reminder interval.
    - e.g. hourly, or every 15 minutes, or randomly between 30 minutes and an hour
    - periodic intervals are aligned from the top of the hour
- Edit or add your own reminder messages.
    - Choose from a set of default bells, or provide your own.
- Separate controls for 'mute' and 'vibrate'.
- Define a time range for daily quiet hours.


## Project Information and Source Code

The source code is available on github at https://github.com/kmac/mindfulnotifier
Please see this link for additional help, or to report any issues.

### Notes on Application Settings in Android

#### Battery Optimization

Some phones will kill apps when running in the background after a relatively
short period of time.  Unfortunately, if your phone is killing the app then you
will stop getting notifications.  If this is happening, you must disable any
'battery optimization' settings for this app. This varies from phone-to-phone -
generally you must go into the settings for the app on your phone and turn off
any battery optimization. See https://dontkillmyapp.com/ for your phone type
and android version for more information.


### Bells

The following bells are included:

- Bell 1: is from user 'steaq' via [https://freesound.org/people/steaq/sounds/346328/](https://freesound.org/people/steaq/sounds/346328/)
and is licensed under the [Creative Commons 0 License][CC0].
- Bell 2: is from user 'itsallhappening' via [https://freesound.org/people/itsallhappening/sounds/48795](https://freesound.org/s/48795/)
and is licensed under the [Creative Commons Sampling+ License][CCS]
- Bell 3: is from user 'suburban grilla' via [https://freesound.org/people/suburban%20grilla/sounds/2166](https://freesound.org/s/2166/)
and is licensed under the [Creative Commons Sampling+ License][CCS]
- Bell 4: is from user 'kerri' via [https://freesound.org/people/kerri/sounds/27421/](https://freesound.org/people/kerri/sounds/27421/)
and is licensed under the [Creative Commons Attribution License][CCA]
- Bell 5: is from user 'dobroide' via [https://freesound.org/people/dobroide/sounds/436976/](https://freesound.org/people/dobroide/sounds/436976/)
and is licensed under the [Creative Commons Attribution License][CCA]

[CC0]: http://creativecommons.org/publicdomain/zero/1.0/   "Creative Commons 0 License"
[CCS]: http://creativecommons.org/licenses/sampling+/1.0/  "Creative Commons Sampling+ License"
[CCA]: https://creativecommons.org/licenses/by/3.0/        "Creative Commons Attribution License"


## License

This project is licensed under the terms of the [GNU General Public License v3.0](https://www.gnu.org/licenses/gpl-3.0.en.html).
""";

Future<void> launchInBrowser(String url) async {
  Uri uri = Uri.parse(url);
  if (await canLaunchUrl(uri)) {
    await launchUrl(uri, mode: LaunchMode.platformDefault);
  } else {
    throw 'Could not launch $url';
  }
}

handleMarkdownOnTap(String text, String href, String title) async {
  if (href != null) {
    logger.i("Launching text=$text, href=$href, title=$title");
    await launchInBrowser(href);
  }
}

class AppAboutListTile extends AboutListTile {
  AppAboutListTile(BuildContext context, {Key key})
      : super(
            key: key,
            icon: Icon(Icons.info),
            applicationVersion: appVersion,
            applicationLegalese:
                'This is open-source free/libre software, licensed under GPL v3.0',
            aboutBoxChildren: [
              Center(
                child: OutlinedButton(
                  child: Text("Visit project on github"),
                  // visualDensity: VisualDensity.compact,
                  onPressed: () {
                    launchInBrowser('https://github.com/kmac/mindfulnotifier');
                  },
                ),
              ),
              Divider(),
              MarkdownBody(
                selectable: true,
                shrinkWrap: true,
                onTapLink: (text, href, title) => handleMarkdownOnTap,
                data: markdownData,
                // styleSheet: MarkdownStyleSheet.fromTheme(Theme.of(context))
                //     .copyWith(
                //         textAlign: WrapAlignment.center,
                //         blockquoteAlign: WrapAlignment.end,
                //         unorderedListAlign: WrapAlignment.center,
                //         listIndent: 2),
              ),
            ]);
}
